CONTRIBUTING.md

```diff
@@ -83,12 +83,17 @@ changes to the API (e.g., remove a public method) or to the ABI (e.g., change re
 * The project aims to provide a consistent experience across all the public APIs. It is important to ensure consistency (same look and feel) across different public packages.
 * Use `final` for public classes everywhere it is possible, this ensures that these classes cannot be extended when the API does not intend to offer that functionality.
 * In general, we use the following ordering of class members:
-    * static fields (final before non-final)
-    * non-static fields (final before non-final)
-    * constructors
-    * static methods
-    * instance methods
-    * inner classes
+    * Static fields (final before non-final)
+    * Instance fields (final before non-final)
+    * Static construction methods (includes static factory methods and builder construction methods)
+    * Constructors
```

Comment:
My reasoning is: the constructor is where the fields get their values. I want the logic close to the fields themselves, so it can be obvious when looking at the class how things are initialized. Static constructor methods should always call actual constructors, and are just helpers to facilitate the actual assignment of the values.

I want to see the assignment logic close to the fields that are being assigned, so I don't have to page through a bunch of static methods to find where that assignment logic might live.

Reply:
Hence static methods above instance fields being the best :P

Reply:
ok, I removed line about "static construction methods", though I find the static construction methods tend to be more interesting than the constructors (when they are both available, the constructors tend to just be to dump params into fields), also static construction methods come first in the flow, calling constructors, another reason I like them above constructors

Reply:
> Hence static methods above instance fields being the best :P

sorry @anuraaga I haven't lived in any codebases that do this, so can't support this (yet) 😂

Reply:
I feel as if both of you prove my point :P

Static factories are the entry point to the API so having them towards the top gives better discoverability within the code. Constructors are private and don't really "deserve" to be higher.

Static fields may be used in static methods so having them even higher than static methods shows what's available to a static method.

Constructors operate on instance fields, grouping instance fields and constructor presents them all together.

All of this is accomplished by having static methods above instance fields.

I heard one counter argument of wanting to group all the state together - I'd argue that this is also satisfied by the proposal. A class is a class, an instance is an instance, they're two separate things so having separate groups for the state seems to just make sense. The instance fields and methods could technically be part of another class even, in which case the grouping becomes syntactically required, and ends up being consistent with when we have static factories in an interface vs a class, we generally don't group all static / non static fields together in practice - I feel as if Java is backing me up too in a way :)

Reply:
So, this is obviously 100% stylistic. Why are we trying to change what we have in place?

Reply:
I think the reason is that instrumentation repo hadn't defined an order yet and we ended up defining it with the static factories above

https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/contributing/style-guideline.md#ordering-of-class-contents

Seeing @trask's interest in changing the ordering, I thought I could also use the chance to push my idea :P But since I already decided to disagree and commit when first talking about this, it's mostly on @trask to fix if we do anything.